File: windows/src/OpenShrimp.Tray/Core/ConfigPeek.cs
namespace OpenShrimp.Tray.Core;

/// <summary>
/// Reads the one config value the tray needs before any core is running: the
/// instance name, which scopes the control endpoint and the scheduled task.
///
/// Deliberately not a YAML parser. Everything else the tray needs about the
/// config it gets from the core over the control channel, and everything it
/// writes goes through "openshrimp config write" — so the schema stays in one
/// language. This reads a single top-level scalar and nothing more.
/// </summary>
internal static class ConfigPeek
{
    public static string? ReadInstanceName(string configPath)
    {
        try
        {
            if (!File.Exists(configPath)) return null;

            foreach (var raw in File.ReadLines(configPath))
            {
                var line = raw.TrimEnd();
                // Top-level keys only: an indented instance_name belongs to
                // some nested mapping and is not the one we mean.
                if (line.Length == 0 || char.IsWhiteSpace(raw[0]) || line.StartsWith('#')) continue;
                if (!line.StartsWith("instance_name:", StringComparison.Ordinal)) continue;

                var value = line["instance_name:".Length..].Trim();
                var comment = value.IndexOf('#');
                if (comment >= 0) value = value[..comment].Trim();
                value = value.Trim('"', '\'');

                return string.IsNullOrEmpty(value) || value == "null" || value == "~" ? null : value;
            }
        }
        catch (IOException)
        {
            // An unreadable config is the default instance's problem to
            // report, not something to fail the tray over.
        }
        return null;
    }
}

File: macos/Sources/OpenShrimpCore/ConfigPeek.swift
import Foundation

/// Reads the one config value the app needs before any core is running: the
/// instance name, which scopes the control endpoint and the login item.
///
/// Deliberately not a YAML parser.  Everything else the app needs about the
/// config it gets from the core over the control channel, and everything it
/// writes goes through `openshrimp config write` — so the schema stays in one
/// language.  This reads a single top-level scalar and nothing more.
enum ConfigPeek {
    static func readInstanceName(at path: String) -> String? {
        guard let contents = try? String(contentsOfFile: path, encoding: .utf8) else {
            // An unreadable config is the default instance's problem to report,
            // not something to fail the app over.
            return nil
        }

        for raw in contents.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = raw.trimmingCharacters(in: CharacterSet(charactersIn: " \t\r"))
            // Top-level keys only: an indented instance_name belongs to some
            // nested mapping and is not the one we mean.
            guard !line.isEmpty, !line.hasPrefix("#"),
                  let first = raw.first, !first.isWhitespace else { continue }
            guard line.hasPrefix("instance_name:") else { continue }

            var value = String(line.dropFirst("instance_name:".count))
                .trimmingCharacters(in: .whitespaces)
            if let comment = value.firstIndex(of: "#") {
                value = String(value[value.startIndex..<comment])
                    .trimmingCharacters(in: .whitespaces)
            }
            value = value.trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))

            return (value.isEmpty || value == "null" || value == "~") ? nil : value
        }
        return nil
    }
}
